src/client-side-encryption/autoEncrypter.ts

```diff
@@ -0,0 +1,641 @@
+          )
+        );
+        return;
+      }
+
+      callback(err, res);
+    };
+
+    if (this._mongocryptdManager.bypassSpawn) {
+      this._mongocryptdClient.connect().then(
+        result => {
+          return _callback(undefined, result);
+        },
+        error => {
+          _callback(error, undefined);
+        }
+      );
+      return;
+    }
+
+    this._mongocryptdManager.spawn(() => {
+      if (!this._mongocryptdClient) {
+        return callback(
+          new MongoRuntimeError(
+            'Reached impossible state: mongocryptdClient is undefined after spawning libmongocrypt.'
+          )
+        );
+      }
+      this._mongocryptdClient.connect().then(
+        result => {
+          return _callback(undefined, result);
+        },
+        error => {
+          _callback(error, undefined);
+        }
+      );
+    });
+  }
+
+  /**
+   * Cleans up the `_mongocryptdClient`, if present.
+   */
+  teardown(force: boolean, callback: Callback<void>) {
+    if (this._mongocryptdClient) {
+      this._mongocryptdClient.close(force).then(
+        result => {
+          return callback(undefined, result);
+        },
+        error => {
+          callback(error);
+        }
+      );
+    } else {
+      callback();
+    }
+  }
+
+  encrypt(ns: string, cmd: Document, callback: Callback<Document | Uint8Array>): void;
+  encrypt(
+    ns: string,
+    cmd: Document,
+    options: CommandOptions,
+    callback: Callback<Document | Uint8Array>
+  ): void;
+  /**
+   * Encrypt a command for a given namespace.
+   */
+  encrypt(
+    ns: string,
+    cmd: Document,
+    options?: CommandOptions | Callback<Document | Uint8Array>,
+    callback?: Callback<Document | Uint8Array>
+  ) {
+    if (typeof ns !== 'string') {
+      throw new TypeError('Parameter `ns` must be a string');
+    }
+
+    if (typeof cmd !== 'object') {
+      throw new TypeError('Parameter `cmd` must be an object');
+    }
+
+    callback = typeof options === 'function' ? options : callback;
+
+    if (callback == null) {
+      throw new TypeError('Callback must be provided');
+    }
+
+    options = typeof options === 'function' ? {} : options;
+
+    // If `bypassAutoEncryption` has been specified, don't encrypt
+    if (this._bypassEncryption) {
+      callback(undefined, cmd);
+      return;
+    }
+
+    const commandBuffer = Buffer.isBuffer(cmd) ? cmd : serialize(cmd, options);
+
+    let context;
+    try {
+      context = this._mongocrypt.makeEncryptionContext(
+        MongoDBCollectionNamespace.fromString(ns).db,
+        commandBuffer
+      );
+    } catch (err) {
+      callback(err, undefined);
+      return;
+    }
+
+    // TODO: should these be accessors from the addon?
+    context.id = this._contextCounter++;
+    context.ns = ns;
+    context.document = cmd;
+
+    const stateMachine = new StateMachine({
+      ...options,
```

Comment:
perhaps questionable if this is needed .. I think we only need the user's BSON serialization options (if that) but future work can investigate this further.

Reply:
The types declared allow `CommandOptions` because that's currently what we pass in, but we only should need the BSON serialization options (and even those are questionable).

Reply:
I think this actually the source of that old sessions bug NODE-4896, if we didn't spread here the session wouldn't have made it to the operations. Do we want to make a change here or file a follow up to investigate?

Reply:
No, the root cause was explicitly passing the session into driver helpers in the StateMachine: https://github.com/mongodb/libmongocrypt/commit/c9e44e7c0842683659e2a068696e6a712c33ab43

I checked in the `StateMachine` and the only four values we're accessing on the options object are the values we pass in here, so I removed the spreading.